### DISCUSSION TITLE
Potentially thread local pollution by MDC

### DISCUSSION THREAD

**Discussion by dgroup** [dgroup/arch4u-pmd#88] [Ideas]
### Summary
| Key            |  Value |
|:-----------|:------|
|  Rule           | PotentiallyThreadLocalPollutionByMdc   |
|Ruleset        | [io/github/dgroup/arch4u/pmd/json/arch4u-logging-ruleset.xml](https://github.com/dgroup/arch4u-pmd/blob/master/src/main/resources/io/github/dgroup/arch4u/pmd/logging/arch4u-logging-ruleset.xml)|
|Category     | Error Prone |
|Framework | [org.slf4j:slf4j-api](https://mvnrepository.com/artifact/org.slf4j/slf4j-api), [org.apache.logging.log4j:log4j-api](https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-api) |
|Since           |    0.1.0    |
|Incidents in past at my experience | 4 |

### Why?
The MDC usage requires a deep understanding of how it works and gives additional side effects that might not be transparent for regular java developers
```java
public ... process(Order order) throws ... {
   MDC.put("orderId", order.getId());
   MDC.put("customer", order.getCustomerId());
   // ... business logic without cleaning MDC thread-local values     
}
```
 thus, MDC thread-local variables should be cleaned within the same method
```java
public ... process(Order order) throws ... {
   try {
         MDC.put("orderId", order.getId());
         MDC.put("customer", order.getCustomerId());      
         // ... business logic 
         this.otherModuleOrOpenSourceLib1.invoke(order);         => each also writes own logs
          \- this.otherModuleOrOpenSourceLib2.invoke(order);    => each also writes own logs 
          \- this.otherModuleOrOpenSourceLib4.invoke(order);    => each also writes own logs
             \- this.otherModuleOrOpenSourceLib5.invoke(order); => each also writes own logs     
             \- this.otherModuleOrOpenSourceLib6.invoke(order); => each also writes own logs
          \- this.otherModuleOrOpenSourceLib4.invoke(order);    => each also writes own logs
       
         this.otherModuleOrOpenSourceLib7.invoke(order);         => each also writes own logs
          \- this.otherModuleOrOpenSourceLib8.invoke(order);    => each also writes own logs  
          \- this.otherModuleOrOpenSourceLib9.invoke(order);    => each also writes own logs
             \- this.otherModuleOrOpenSourceLib5.invoke(order); => each also writes own logs     
             \- this.otherModuleOrOpenSourceLib6.invoke(order); => each also writes own logs
          \- this.otherModuleOrOpenSourceLib10.invoke(order);   => each also writes own logs
   } finally {
       MDC.remove("orderId");
       MDC.remove("customer");
   }
   ...
}
```
Even, if people know how MDC works/are aware, they might forget to clean in another place in a project or due to staff turnover (which is common in our industry) newcomers might not aware and may make mistakes.

**Comment by Cimballi** [comment]
Wouldn't it be better to handle MDC cleanup in filter / interceptor classes, in order to avoid repeating those finally blocks ?
